command/debug.go

```diff
@@ -0,0 +1,945 @@
+		c.flagOutput = fmt.Sprintf("vault-debug-%s", formattedTime)
+	}
+
+	// Strip trailing slash before proceeding
+	c.flagOutput = strings.TrimSuffix(c.flagOutput, "/")
+
+	// If compression is enabled, trim the extension so that the files are
+	// written to a directory even if compression somehow fails. We ensure the
+	// extension during compression. We also prevent overwriting if the file
+	// already exists.
+	dstOutputFile := c.flagOutput
+	if c.flagCompress {
+		if !strings.HasSuffix(dstOutputFile, ".tar.gz") && !strings.HasSuffix(dstOutputFile, ".tgz") {
+			dstOutputFile = dstOutputFile + debugCompressionExt
+		}
+
+		// Ensure that the file doesn't already exist, and ensure that we always
+		// trim the extension from flagOutput since we'll be progressively
+		// writing to that.
+		if _, err := os.Stat(dstOutputFile); os.IsNotExist(err) {
+			c.flagOutput = strings.TrimSuffix(c.flagOutput, ".tar.gz")
+			c.flagOutput = strings.TrimSuffix(c.flagOutput, ".tgz")
+		} else {
+			return nil, "", fmt.Errorf("output file already exists: %s", dstOutputFile)
+		}
+	}
+
+	// Stat check the directory to ensure we don't override any existing data.
+	if _, err := os.Stat(c.flagOutput); os.IsNotExist(err) {
+		err := os.MkdirAll(c.flagOutput, 0755)
+		if err != nil {
+			return nil, "", fmt.Errorf("unable to create output directory: %s", err)
+		}
+	} else {
+		return nil, "", fmt.Errorf("output directory already exists: %s", c.flagOutput)
+	}
+
+	// Instantiate the channels for polling
+	c.metricsCh = make(chan map[string]interface{})
+	c.serverStatusCh = make(chan map[string]interface{})
+	c.replicationStatusCh = make(chan map[string]interface{})
+	c.hostInfoCh = make(chan map[string]interface{})
+
+	// Populate initial index fields
+	c.debugIndex = &debugIndex{
+		VaultAddress:    client.Address(),
+		ClientVersion:   version.GetVersion().VersionNumber(),
+		Compress:        c.flagCompress,
+		Duration:        int(c.flagDuration.Seconds()),
+		Interval:        int(c.flagInterval.Seconds()),
+		MetricsInterval: int(c.flagMetricsInterval.Seconds()),
+		RawArgs:         rawArgs,
+		Version:         debugIndexVersion,
+		Targets:         c.flagTargets,
+		Timestamp:       captureTime,
+		Errors:          []*captureError{},
+	}
+
+	return client, dstOutputFile, nil
+}
+
+func (c *DebugCommand) defaultTargets() []string {
+	return []string{"config", "host", "metrics", "pprof", "replication-status", "server-status"}
+}
+
+func (c *DebugCommand) captureStaticTargets(client *api.Client) error {
+	// Capture configuration state
+	if strutil.StrListContains(c.flagTargets, "config") {
+		c.logger.Info("capturing configuration state")
+
+		resp, err := client.Logical().Read("sys/config/state/sanitized")
+		if err != nil {
+			captErr := newCaptureError("config", err)
+			c.debugIndex.Errors = append(c.debugIndex.Errors, captErr)
+
+			c.logger.Error("config: error capturing config state", "error", err)
+		}
+
+		if resp != nil && resp.Data != nil {
```

Comment:
Should it be an error to receive no data in the response?

Reply:
I don't think that would be strictly an error, we simply don't append anything to the collection.

Reply:
Why wouldn't it be an error?  The user asked for config data, so we asked Vault for the config data, and it gave us none.  Sounds like an error to me, even if the Vault server didn't report one.